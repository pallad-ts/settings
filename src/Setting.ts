import {Maybe, Validation} from "monet";

const DEFAULT_VALIDATOR: Setting.Validator<any> = (value: unknown) => Validation.Success(value);

export abstract class Setting<TType> {
	constructor(protected defaultValue: TType,
				protected validator: Setting.Validator<TType> = DEFAULT_VALIDATOR) {

	}

	abstract merge(values: Maybe<unknown[]>): Promise<Validation<string, TType>>;
}

export namespace Setting {
	export interface Validator<T> {
		(value: unknown): Promise<Validator.Result<T>> | Validator.Result<T>;
	}

	export namespace Validator {
		export type Result<T> = Validation<string, T>;
	}

	export type Type<T extends Setting<any>> = T extends Setting<infer TType> ? TType : never;
}
