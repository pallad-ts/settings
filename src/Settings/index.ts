export * from './ObjectSetting';
export * from './PrimitiveSetting';
