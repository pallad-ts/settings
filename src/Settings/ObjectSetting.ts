import {Maybe, Validation} from "monet";
import {Setting} from "../Setting";

export class ObjectSetting<TType extends Record<any, any>> extends Setting<TType> {
	async merge(values: Maybe<unknown[]>): Promise<Setting.Validator.Result<TType>> {
		if (values.isNone()) {
			return Validation.Success(this.defaultValue);
		}

		let currentValue = {
			...this.defaultValue
		};

		for (const override of values.some()) {
			const validationResult = await this.validator({
				...currentValue,
				...override as {}
			});
			if (validationResult.isFail()) {
				return validationResult;
			}
			currentValue = validationResult.success();
		}
		return Validation.Success(currentValue);
	}
}
