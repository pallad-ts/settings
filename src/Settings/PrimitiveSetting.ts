import {Maybe, Validation} from "monet";
import {Setting} from "../Setting";

export class PrimitiveSetting<TType> extends Setting<TType> {
	async merge(values: Maybe<unknown[]>): Promise<Setting.Validator.Result<TType>> {
		if (values.isNone()) {
			return Validation.Success(this.defaultValue);
		}

		const lastValue = values.some()[values.some().length - 1];
		return this.validator(lastValue);
	}
}
