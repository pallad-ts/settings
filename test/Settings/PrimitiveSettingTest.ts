import {Setting} from "@src/Setting";
import {Maybe, Validation} from "monet";
import * as sinon from 'sinon';
import {PrimitiveSetting} from "@src/Settings";

describe('PrimitiveSetting', () => {
	const DEFAULT_VALUE = 'SomeDefaultValue';
	describe('merging', () => {
		const setting = new PrimitiveSetting(DEFAULT_VALUE);

		it('returns default value if no override provided', () => {
			return expect(setting.merge(Maybe.None()))
				.resolves
				.toEqual(Validation.Success(DEFAULT_VALUE));
		});

		it('returns last value if override provided', () => {
			return expect(setting.merge(Maybe.Some(['foo', 'bar', 'zee'])))
				.resolves
				.toEqual(Validation.Success('zee'));
		});

		it('fails if last values is invalid', async () => {
			const validator = sinon.stub().callsFake(value => {
				if (value === 'zee') {
					return Validation.Fail('Invalid value');
				}
				return Validation.Success(value);
			})
			const setting = new PrimitiveSetting(DEFAULT_VALUE, validator)
			await expect(setting.merge(Maybe.Some(['foo', 'bar', 'zee'])))
				.resolves
				.toMatchSnapshot();
			sinon.assert.calledOnce(validator);
		});

		it('returns value sanitized by validation', async () => {
			const validator = sinon.stub().callsFake(value => {
				return Validation.Success(value.toUpperCase());
			})
			const setting = new PrimitiveSetting(DEFAULT_VALUE, validator)
			await expect(setting.merge(Maybe.Some(['foo', 'bar', 'zee'])))
				.resolves
				.toEqual(Validation.Success('ZEE'));
			sinon.assert.calledOnce(validator);
		});
	});
});
