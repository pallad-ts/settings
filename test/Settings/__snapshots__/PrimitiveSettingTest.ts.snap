// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimitiveSetting merging fails if last values is invalid 1`] = `
Object {
  "@@type": "monet.js/Validation",
  "isSuccessValue": false,
  "val": "Invalid value",
}
`;
